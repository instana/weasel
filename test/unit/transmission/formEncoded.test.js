jest.mock('../../../lib/browser');
jest.mock('../../../lib/vars');

describe('transmission/formEncoded', () => {
  let browserMock;
  let sendBeacon;

  beforeEach(() => {
    browserMock = require('../../../lib/browser');
    browserMock.reset();
    sendBeacon = require('../../../lib/transmission/formEncoded').sendBeacon;
  });

  it('must do nothing when there is nothing to transmit', () => {
    sendBeacon({});
    assertMatchesSnapshot();
  });

  it('must only transmit own props', () => {
    const parentObject = {};
    parentObject.shouldNotBeIncluded = 'secr3t!';
    const beacon = Object.create(parentObject);
    beacon.page = 'cart';
    sendBeacon(beacon);
    assertMatchesSnapshot();
  });

  it('must transmit small payloads via image request', () => {
    sendBeacon({
      page: 'cart',
      id: 42,
      m_featureFlags: JSON.stringify({developerMode: false})
    });
    assertMatchesSnapshot();
  });

  it('must transmit larger payloads via XHR request', () => {
    const beacon = {
      page: 'cart',
      id: 42,
      m_featureFlags: JSON.stringify({developerMode: false}),
      longKey: Array(2001).fill('a').join('')
    };

    sendBeacon(beacon);
    assertMatchesSnapshot();
  });

  function assertMatchesSnapshot() {
    expect({
      imageRequests: browserMock.imageRequests,
      xhrRequests: browserMock.xhrRequests
    }).toMatchSnapshot();
  }
});
