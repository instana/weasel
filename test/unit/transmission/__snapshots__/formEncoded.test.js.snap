// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmission/formEncoded must do nothing when there is nothing to transmit 1`] = `
Object {
  "imageRequests": Array [],
  "xhrRequests": Array [],
}
`;

exports[`transmission/formEncoded must only transmit own props 1`] = `
Object {
  "imageRequests": Array [
    "http://eum.example.com?page=cart",
  ],
  "xhrRequests": Array [],
}
`;

exports[`transmission/formEncoded must transmit larger payloads via XHR request 1`] = `
Object {
  "imageRequests": Array [],
  "xhrRequests": Array [
    Object {
      "__secret__": Object {
        "ignored": true,
      },
      "open": Array [
        "POST",
        "http://eum.example.com",
        true,
      ],
      "requestHeader": Array [
        Array [
          "Content-type",
          "application/x-www-form-urlencoded;charset=UTF-8",
        ],
      ],
      "responseType": "text",
      "send": Array [
        "page=cart&id=42&m_featureFlags=%7B%22developerMode%22%3Afalse%7D&longKey=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
      ],
      "timeout": 5432,
    },
  ],
}
`;

exports[`transmission/formEncoded must transmit small payloads via image request 1`] = `
Object {
  "imageRequests": Array [
    "http://eum.example.com?page=cart&id=42&m_featureFlags=%7B%22developerMode%22%3Afalse%7D",
  ],
  "xhrRequests": Array [],
}
`;
